compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala

```diff
@@ -230,7 +230,8 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
 
     allNonLocalClassesInSrc += cl
 
-    if (sym.isStatic && ctx.platform.hasMainMethod(sym)) {
+    if (sym.isStatic && !sym.is(Trait) && ctx.platform.hasMainMethod(sym)) {
+       // If sym is an object, all main methods count, otherwise only @static ones count.
```

Comment:
Should this comment be moved to `hasMainMethod` where the test is performed?

Reply:
There's another caller of hasMainMethod in the backend. Not knowing what
this does or expects, I am not sure whether the change
is OK or not.

Reply:
So the other caller is https://github.com/lampepfl/dotty/blob/95ae77c4748bd105de45601875b447ee1fe6a98b/compiler/src/dotty/tools/dotc/transform/CollectEntryPoints.scala#L66-L67
which will emit a warning if there's a main method in an object and also a main method in the companion class of the object, and we do want the warning even if the main method in the companion class is not static, since its mere presence means that the backend will not generate a static forwarder from the main in the object to the companion class.

Reply:
The situation is very confusing because there's two source files called CollectEntryPoints and none of them actually work, I've made a separate PR to cleanup things a bit: https://github.com/lampepfl/dotty/pull/6981/commits/89801f96a689b7626a99b5513c557e788a581717

Reply:
@smarter Do you want to take this over, or should we merge as is? I don't have the time to do more work on this.